// @edgio/core/runtime/edge-functions/EdgeFunctionMetrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class EdgeFunctionMetrics {
    constructor() {
        this.activeTimers = {};
        this.nextActiveTimerId = 0;
        this.values = {};
        // Connascence of value with Sailfish. See wasm_metrics.h
        this.FIRST_INDEX = 0;
        this.LAST_INDEX = 9;
    }
    microseconds() {
        const hrTime = process.hrtime();
        return hrTime[0] * 1000000 + hrTime[1] / 1000;
    }
    isValidIndex(index) {
        return index >= this.FIRST_INDEX && index <= this.LAST_INDEX;
    }
    isActiveIndex(index) {
        return !!this.activeTimers[index];
    }
    startTimer(index) {
        if (this.isActiveIndex(index)) {
            return false;
        }
        this.activeTimers[this.nextActiveTimerId] = [index, this.microseconds()];
        return true;
    }
    // NOTE: In Sailfish, it is critical for security that we do not
    // allow the user to access the time or delta times from their
    // EdgeFunction. But in the CLI, this is not a concern.
    stopTimer(index) {
        if (!this.isActiveIndex(index)) {
            return false;
        }
        const startTime = this.activeTimers[index][1];
        this.add(index, this.microseconds() - startTime);
        delete this.activeTimers[index];
        return true;
    }
    add(index, value) {
        this.values[index] = (this.values[index] || 0) + value;
    }
    write(log) {
        const keys = Object.keys(this.values);
        for (let i = 0; i < keys.length; i++) {
            const index = parseInt(keys[i]);
            const value = this.values[index];
            if (value) {
                log(`Metrics: index:${index} value:${value}`);
            }
        }
    }
}
exports.default = EdgeFunctionMetrics;
