// @edgio/core/deploy/edge-functions/EdgeFunctionsBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const edgeFunctionUtils_1 = require("../../utils/edgeFunctionUtils");
/**
 * The abstract base class for edge function builders.
 */
class EdgeFunctionsBuilder {
    /**
     * @param rules The router rules
     */
    constructor(rules) {
        this.functionDefaultExportName = 'edgeFunctionsDefaultExport';
        this.indexCodeTemplateFilename = 'edgeFunctionsIndexCodeTemplate.js';
        this.rules = [];
        this.rules = rules;
    }
    /**
     * Returns the array of function entries.
     * Each entry contains the name of the function and the code of function which will return the edge function.
     * @returns FunctionEntry[]
     */
    getFunctionEntries() {
        const features = (0, edgeFunctionUtils_1.getEdgeFunctionFeatures)(this.rules);
        return features.map(feature => {
            const srcPath = feature.edge_function;
            const functionCode = this.getFunctionCode(srcPath);
            return {
                name: srcPath,
                code: `function() {
          ${functionCode}
          return ${this.functionDefaultExportName}
        }`,
            };
        });
    }
    /**
     * Returns the final index code with the edge functions map inserted.
     * @returns
     */
    async getIndexCode() {
        const functionEntries = this.getFunctionEntries();
        const indexCodeTemplate = await this.getIndexCodeTemplate();
        return indexCodeTemplate.replace(/["'`]__INSERT_EDGE_FUNCTIONS_MAP_HERE__["'`]/, `new Map([${functionEntries.map(entry => `['${entry.name}',${entry.code}]`).join(',\n')}])`);
    }
}
exports.default = EdgeFunctionsBuilder;
